### DISCUSSION TITLE
CPU vs GPU comparison

### DISCUSSION THREAD

**Discussion by synjon-symthe** [pybamm-team/PyBaMM#1780] [General]
Hello PyBaMMers:

Does anyone know of a reference where the performance of PyBaMM run on a CPU has been compared with the same problem being run on a GPU?

I tried to modify Tutorial 8 (Tutorial 8 - Solver options.ipynb) so it has 320 gridpoints and ran it on a single cpu (AMD EPYC 7542) and then on a A100, the timing was about the same.

Does PyBaMM have any performance benchmarks like those shown in 
https://openmm.org/benchmarks

Thanks
Synjon

**Comment by valentinsulzer** [comment]
Some benchmarks are here https://pybamm-team.github.io/pybamm-bench/, we should definitely add more

PyBaMM's solvers are definitely not optimized for GPUs. Once #1129 is completed that will become more doable

**Reply by synjon-symthe** [reply]
Thanks, Valentin.  I am not sure what I should be looking at in the pybamm-bench link. I was looking at the benchmark list tab and I see a lot of benchmark names but I don't see the python code that users can test out on their systems.  I don't see any information on the hardware on which the benchmarks were run. Is there some other link which has this information?  The Openmm benchmarks show a comparison of the time to solution for various architectures, I was wondering if there is some information about these PyBaMM benchmarks.

**Reply by valentinsulzer** [reply]
We don't have anything like that set up. Would be great to have though, if you know how to do it :)

**Reply by synjon-symthe** [reply]
Thanks Valentin.  If we have a repo with a set of problems, users can download them and run them on the machines available to them and submit the information.  It will be helpful to the community.
As I said in my earlier note, I merged tutorials 8 and 9 and made the mesh size a variable and ran the case on my mac and also a CPU and GPU.  I have attached a pdf file of the python code (I can't attach a .py file itself).  Here are my results
Mac (2.6 GHz Intel Core i7)
Safe mode solve time: 55.300 s
Fast mode solve time: 20.853 s
Safe mode solve time: 322.278 ms
Fast mode solve time: 118.184 ms

**AMD EPYC 7542 (AMD Rome) - 1.496GHz**
Safe mode solve time: 84.14937464892864
Fast mode solve time: 15.45244575291872
Safe mode solve time: 0.6143575999885798
Fast mode solve time: 0.09564683400094509

Intel(R) Xeon(R) CPU E5-2695 2.10 GHz
Safe mode solve time: 92.90927498787642
Fast mode solve time: 18.225966915953904
Safe mode solve time: 0.5063529568724334
Fast mode solve time: 0.11937079997733235

GPU (A100)
Safe mode solve time: 79.15984221099643
Fast mode solve time: 13.566242505039554
Safe mode solve time: 0.40356054098811
Fast mode solve time: 0.09249094099504873

If there are some example cases I that I could try to run and post timing results I will be glad to do so.  If the problem size can be varied by changing the mesh parameter (as I have done), I can try out some larger cases as well.
[sample1.pdf](https://github.com/pybamm-team/PyBaMM/files/7470364/sample1.pdf)

**Reply by valentinsulzer** [reply]
The code for existing benchmarks is here: https://github.com/pybamm-team/PyBaMM/tree/develop/benchmarks
We can easily add some more.
If you can take the lead on creating a platform for people to report results that would be amazing. I don't think anyone from the pybamm team has time to do this

**Comment by synjon-symthe** [comment]
Valentin:
I am not sure what you mean by "creating a platform for people to report results".  If people are running the two python codes in the benchmarks link, we can allow them to upload their timings with some information on the cpu or gpu core they are using. Won't that be enough?

    On Thursday, November 4, 2021, 11:07:04 AM CDT, Valentin Sulzer ***@***.***> wrote:  
 
 


The code for existing benchmarks is here: https://github.com/pybamm-team/PyBaMM/tree/develop/benchmarks
We can easily add some more.
If you can take the lead on creating a platform for people to report results that would be amazing. I don't think anyone from the pybamm team has time to do this

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by valentinsulzer** [reply]
> we can allow them to upload their timings with some information on the cpu or gpu core they are using

Where do you suggest they upload these timings?

**Reply by synjon-symthe** [reply]
Would it not be possible to have a benchmark tab on the pybamm home-page where users can upload their results. It would be helpful to have select problems on the benchmark page which can be easily downloaded and some blurb on the expected output on successful completion.
I am not quite sure how to run the problem on the benchmark's link you sent (https://github.com/pybamm-team/PyBaMM/tree/develop/benchmarks)
